src/dinit-iostream.cc

```diff
@@ -0,0 +1,744 @@
+#include <limits>
+#include <cstdio>
+#include <cstring>
+
+#include <sys/types.h>
+#include <unistd.h>
+#include <fcntl.h>
+
+#include <dinit-iostream.h>
+#include <baseproc-sys.h>
+
+// See dinit-iostream.h for inforamtion about the usage
+
+namespace {
+
+// A simple trick for calculating how much digits can be in a type with compile-time integer
+// logarithm like function.
+template <typename T> constexpr unsigned type_max_numdigits(T num = std::numeric_limits<T>::max(),
+        unsigned pow = 0)
+{
+    return (num == 0) ? pow : type_max_numdigits(num / 10, pow + 1);
+}
+
+// We should have space for sign in signed variables.
+constexpr unsigned UINT_MAX_CHARS = type_max_numdigits<unsigned>();
+constexpr unsigned INT_MAX_CHARS = type_max_numdigits<int>() + 1;
+constexpr unsigned ULONG_MAX_CHARS = type_max_numdigits<unsigned long>();
+constexpr unsigned LONG_MAX_CHARS = type_max_numdigits<long>() + 1;
+
+} /* anonymous namespace */
+
+namespace dio {
+
+// class io_base
+
+streambuf *io_base::get_buf() noexcept
+{
+    return buf.get();
+}
+
+bool io_base::is_open() noexcept
+{
+    return (fd >= 0);
+}
+
+void io_base::set_fd(const int newfd) noexcept
+{
+    fd = newfd;
+}
+
+int io_base::get_fd() noexcept
+{
+    return fd;
+}
+
+// class ostream
+
+ssize_t ostream::put(const char *msg, size_t count) noexcept
+{
+    if (!good()) {
+        return -1;
+    }
+
+    if (count == 0) {
+        return 0;
+    }
+
+    size_t remaining_bytes = count;
+    while (remaining_bytes > buf->get_free()) {
+        struct iovec bufv[3];
+        char *ptr = buf->get_ptr(0);
+
+        bufv[0].iov_base = ptr;
+        bufv[0].iov_len = buf->get_contiguous_length(ptr);
+        if (buf->get_length() != bufv[0].iov_len) {
+            // buffer's data in streambuf is splited into two parts in this case because of cpbuffer nature.
```

Comment:
"streambuf is _split_ into two parts"

_splitted_ is not really used in English (and anyway it's spelt incorrectly here)